src/nlp/dataset_dict.py

```diff
@@ -0,0 +1,236 @@
+from typing import Dict, List, Optional
+
+import numpy as np
+
+from .features import Features
+
+
+class DatasetDict(dict):
```

Comment:
We could maybe override `__setitem__` to make sure all values are `Dataset` or maybe simpler and more flexible is to have an assert and a clean clean error message in all the processing methods (`map`, `filter` etc) for this case (otherwise it will fail with a not-super-explicit "No map method for ...").